meson.build

```diff
@@ -0,0 +1,174 @@
+# pgBackRest Project
+####################################################################################################################################
+project(
+    'pgbackrest',
+    ['c'],
+    version: '2.40dev',
+    license: 'MIT',
+    meson_version: '>=0.53.2',
+    default_options: [
+        # Core options
+        'buildtype=release',
+        'warning_level=2',
+
+        # Base options
+        'b_ndebug=if-release',
+
+        # Compiler options
+        'c_std=c99',
+    ],
+)
+
+####################################################################################################################################
+# OS-specific settings
+####################################################################################################################################
+if build_machine.system() == 'linux'
+    add_global_arguments('-D_POSIX_C_SOURCE=200809L', language : 'c')
+elif build_machine.system() == 'darwin'
+    add_global_arguments('-D_DARWIN_C_SOURCE', language : 'c')
+endif
+
+####################################################################################################################################
+# Enable as many additional warnings as possible to catch potential errors
+####################################################################################################################################
+warning_enable = [
+    # Warn for implicit conversions that may alter a value
+    'conversion',
+
+    # Warn about duplicated conditions in an if-else-if chain
+    'duplicated-cond',
+
+    # Warn when an if-else has identical branches
+    'duplicated-branches',
+
+    # Warn if the format string is not a string literal and cannot be checked
+    'format-nonliteral',
+
+    # Enable -Wformat plus additional format checks
+    'format=2',
+
+    # Warn if the format string requires an unsigned argument and the argument is signed and vice versa
+    'format-signedness',
+
+    # Warn about anything that depends on the “size of” a function type or of void
+    'pointer-arith',
+
+    # Warn if a function is declared or defined without specifying the argument types
+    'strict-prototypes',
+
+    # Warn if a variable-length array is used
+    'vla',
+
+    # Give string constants the type const char[length] so that copying the address of one into a non-const char * pointer produces
+    # a warning
+    'write-strings',
+]
+
+foreach warning : warning_enable
+    if meson.get_compiler('c').has_argument('-W' + warning)
+        add_project_arguments('-W' + warning, language: 'c')
+    endif
+endforeach
+
+####################################################################################################################################
+# Disable various unhelpful warnings by adding -Wno-warning switches. GCC won't complain about unrecognized -Wno-warning switches,
+# so we have to test for the positive form, and if that works, add the negative form.
```

Comment:
At one point I was trying to target 0.45 and this was introduced in 0.51. For now I've added a note.

I'm waiting back from the Debian packagers to find out if it would be OK to specify a version of meson for the build. Their initial answer was no, but not sure they are aware it is a Python module.

If we must use the meson version that is available in the distro then we either need to figure out how to work with 0.45 or wait a few years to remove autoconf/make.

Reply:
Ubuntu 18.04, https://packages.ubuntu.com/bionic/meson.

We've used backports to get a working version of the YAML Perl module before, so that might be an option.

Reply:
Well, regardless of how I feel about Ubuntu, the apt packagers build for it so it needs to work.

Reply:
So, yeah, I mean Debian-based. :)

Reply:
I decided to leave the code as-is, but added a comment explaining why we are not using the new feature in 49ec515ed8f3b. Look OK?

Reply:
:+1:

Sure. The most important thing is to use whatever you need in order to handle the supported userbase. I generally recommend people either update to newer Meson usage, or add a TODO about the new feature, **whichever the project feels is right**. And what is right for you is going the TODO route, which you have done.